packages/edit-site/src/components/sync-state-with-url/use-init-edited-entity-from-url.js

```diff
@@ -36,60 +37,118 @@ export function useInitEditedEntity( { postId, postType } ) {
 		};
 	}, [] );
 
-	const {
-		setEditedEntity,
-		setTemplate,
-		setTemplatePart,
-		setPage,
-		setNavigationMenu,
-	} = useDispatch( editSiteStore );
+	const resolvedTemplateId = useSelect(
+		( select ) => {
+			// If we're rendering a post type that doesn't have a template
+			// no need to resolve its template.
+			if ( postTypesWithoutParentTemplate.includes( postType ) ) {
+				return undefined;
+			}
 
-	useEffect( () => {
-		if ( postType && postId ) {
-			switch ( postType ) {
-				case TEMPLATE_POST_TYPE:
-					setTemplate( postId );
-					break;
-				case TEMPLATE_PART_POST_TYPE:
-					setTemplatePart( postId );
-					break;
-				case NAVIGATION_POST_TYPE:
-					setNavigationMenu( postId );
-					break;
-				case PATTERN_TYPES.user:
-					setEditedEntity( postType, postId );
-					break;
-				default:
-					setPage( {
-						context: { postType, postId },
-					} );
+			const {
+				getEditedEntityRecord,
+				getEntityRecords,
+				getDefaultTemplateId,
+				__experimentalGetTemplateForLink,
+			} = select( coreDataStore );
+
+			function resolveTemplateForPostTypeAndId(
+				postTypeToResolve,
+				postIdToResolve
+			) {
+				const editedEntity = getEditedEntityRecord(
+					'postType',
+					postTypeToResolve,
+					postIdToResolve
+				);
+				if ( ! editedEntity ) {
+					return undefined;
+				}
+				// First see if the post/page has an assigned template and fetch it.
+				const currentTemplateSlug = editedEntity.template;
+				if ( currentTemplateSlug ) {
+					const currentTemplate = getEntityRecords(
+						'postType',
+						'wp_template',
+						{
+							per_page: -1,
+						}
+					)?.find( ( { slug } ) => slug === currentTemplateSlug );
+					if ( currentTemplate ) {
+						return currentTemplate.id;
+					}
+				}
+
+				// If the no template is assigned, use the default template.
+				return getDefaultTemplateId( {
+					slug: `${ postTypeToResolve }-${ editedEntity?.slug }`,
+				} );
+			}
+
+			// If we're rendering a specific page, post... we need to resolve its template.
+			if ( postType && postId ) {
+				return resolveTemplateForPostTypeAndId( postType, postId );
+			}
+
+			// If we're rendering the home page, and we have a static home page, resolve its template.
+			if ( homepageId ) {
+				return resolveTemplateForPostTypeAndId( 'page', homepageId );
 			}
 
-			return;
+			// If we're not rendering a specific page, use the front page template.
+			if ( ! isRequestingSite && url ) {
+				const template = __experimentalGetTemplateForLink( url );
+				return template?.id;
+			}
+		},
+		[ homepageId, isRequestingSite, url, postId, postType ]
+	);
+
+	// todo add template slug in context.
```

Comment:
I noticed that in some situations we store the templateSlug in the context, I'm not entirely sure the reason for this. I'll debug.